transport/grpc/transport.go

```diff
@@ -95,36 +96,52 @@ func (t *Transport) NewOutbound(peerChooser peer.Chooser, options ...OutboundOpt
 }
 
 // RetainPeer retains the peer.
-func (t *Transport) RetainPeer(peerIdentifier peer.Identifier, peerSubscriber peer.Subscriber) (peer.Peer, error) {
+//
+// Deprecated: use grpcTransport.NewDialer(...grpc.DialOption) to create a
+// peer.Transport instead of using the grpc.Transport as a peer.Transport.
+// You will need to explicitly choose either grpc.WithInsecure() or
+// grpc.WithCredentials(...) options.
+func (t *Transport) RetainPeer(pid peer.Identifier, ps peer.Subscriber) (peer.Peer, error) {
+	return t.retainPeer(pid, nil, ps)
+}
+
+// retainPeer provides the general implementation of retainPeer for both TLS
+// and non-TLS clients.
+func (t *Transport) retainPeer(pid peer.Identifier, options *dialOptions, ps peer.Subscriber) (peer.Peer, error) {
 	t.lock.Lock()
 	defer t.lock.Unlock()
-	address := peerIdentifier.Identifier()
+	address := pid.Identifier()
 	p, ok := t.addressToPeer[address]
 	if !ok {
 		var err error
-		p, err = newPeer(address, t)
+		p, err = t.newPeer(address, options)
 		if err != nil {
 			return nil, err
 		}
 		t.addressToPeer[address] = p
 	}
-	p.Subscribe(peerSubscriber)
+	p.Subscribe(ps)
 	return p, nil
 }
 
 // ReleasePeer releases the peer.
-func (t *Transport) ReleasePeer(peerIdentifier peer.Identifier, peerSubscriber peer.Subscriber) error {
+//
+// Deprecated: use grpcTransport.NewDialer(...grpc.DialOption) to create a
+// peer.Transport instead of using the grpc.Transport as a peer.Transport.
+// You will need to explicitly choose either grpc.WithInsecure() or
+// grpc.WithCredentials(...) options.
+func (t *Transport) ReleasePeer(pid peer.Identifier, ps peer.Subscriber) error {
 	t.lock.Lock()
 	defer t.lock.Unlock()
-	address := peerIdentifier.Identifier()
+	address := pid.Identifier()
```

Comment:
nit: Does pid.Identifier() need to be protected by the lock? It seems external to Transport

Reply:
The identifier is not mutable, so it is not necessary for it to be accessed under the lock. There is an almost negligible optimization possible here, but I’d like to be conservative and keep the lock preamble on top.